contrib/win32/win32compat/tncon.c

```diff
@@ -148,639 +150,676 @@ ReadConsoleForTermEmul(HANDLE hInput, char *destin, int destinlen)
-							((dwControlKeyState & LEFT_CTRL_PRESSED) || (dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)ALT_CTRL_PF7_KEY, strlen(ALT_CTRL_PF7_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & RIGHT_ALT_PRESSED) ||
-							(dwControlKeyState & LEFT_ALT_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_ALT_PF7_KEY, strlen(SHIFT_ALT_PF7_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & LEFT_CTRL_PRESSED) ||
-							(dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_CTRL_PF7_KEY, strlen(SHIFT_CTRL_PF7_KEY), 0);
-						break;
-					case VK_F8:
-						if (dwControlKeyState == 0)
-							NetWriteString2(pParams->Socket, (char *)PF8_KEY, strlen(PF8_KEY), 0);
-
-						else if (dwControlKeyState == SHIFT_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)SHIFT_PF8_KEY, strlen(SHIFT_PF8_KEY), 0);
-
-						else if (dwControlKeyState == LEFT_CTRL_PRESSED || dwControlKeyState == RIGHT_CTRL_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)CTRL_PF8_KEY, strlen(CTRL_PF8_KEY), 0);
-
-						else if (dwControlKeyState == LEFT_ALT_PRESSED || dwControlKeyState == RIGHT_ALT_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)ALT_PF8_KEY, strlen(ALT_PF8_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & RIGHT_ALT_PRESSED) ||
-							(dwControlKeyState & LEFT_ALT_PRESSED)) && ((dwControlKeyState & LEFT_CTRL_PRESSED) ||
-							(dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_ALT_CTRL_PF8_KEY, strlen(SHIFT_ALT_CTRL_PF8_KEY), 0);
+		ReadConsoleInputW(hInput, inputRecordArray, inputRecordArraySize, &dwInput);
+
+		for (DWORD i=0; i < dwInput; i++) {
+			INPUT_RECORD inputRecord = inputRecordArray[i];
+
+			switch (inputRecord.EventType) {
+			case WINDOW_BUFFER_SIZE_EVENT:
+				queue_terminal_window_change_event();
+				break;
+
+			case FOCUS_EVENT:
+				/* FALLTHROUGH */
+			case MENU_EVENT:
+				break;
+
+			case KEY_EVENT:
+				if (((inputRecord.Event.KeyEvent.bKeyDown) || (!inputRecord.Event.KeyEvent.bKeyDown && inputRecord.Event.KeyEvent.wVirtualKeyCode == VK_MENU)) &&
+				    (inputRecord.Event.KeyEvent.uChar.UnicodeChar != L'\0')) {
+					int n = WideCharToMultiByte(
+						CP_UTF8,
+						0,
+						&(inputRecord.Event.KeyEvent.uChar.UnicodeChar),
+						1,
+						(LPSTR)octets,
+						20,
+						NULL,
+						NULL);
+
+					if (IS_HIGH_SURROGATE(inputRecord.Event.KeyEvent.uChar.UnicodeChar)) {
+						utf16_surrogatepair[0] = inputRecord.Event.KeyEvent.uChar.UnicodeChar;
+						break; // break to read low surrogate.
+					}
+					else if (IS_LOW_SURROGATE(inputRecord.Event.KeyEvent.uChar.UnicodeChar)) {
+						utf16_surrogatepair[1] = inputRecord.Event.KeyEvent.uChar.UnicodeChar;
+					}
```

Comment:
nit: this doesn't handle what happens if you get a high surrogate and then a character that is not a surrogate pair. To be fair, this may not matter.

Reply:
I don't think it's possible? If yes then it will be a bug on the console win32 API.

Reply:
Users can put whatever they want on their clipboards. Console doesn't control that.

Reply:
How can users break the surrogate pair? Isn't the surrogate pair generated based on the keystroke i.e., it can't be broken apart?

Reply:
Contents on the clipboard are not generated by user keystrokes. They can be inserted by any application anywhere at any time. The clipboard does not map to keystrokes.

Reply:
The current code does handle this scenario (having any number of characters between a surrogate pair)..

Reply:
It is incorrect to handle a broken surrogate pair in this way.

This is what is happening-

If your input stream looks like this

```
SURR    A    A    A    A SURR
D83D 0041 0041 0041 0041 DD74
```


the _only_ valid way to handle that input is:

```
invalid character
A
A
A
invalid character
```

Right now, your code will do this:

```
A
A
A
🕴
```

That is invalid. The surrogate pair was broken. It was broken into two different characters.

Reply:
You confuse are confusing me. Your first comment says we should handle any character between surrogate pair. Now you are contradicting it.

Reply:
I'm sorry I was not clear. My initial comment was that this code does "not handle what happens if you get a high surrogate and then any other character", but I didn't specify the correct handling for this case.

The correct way to handle a surrogate pair is the following.

If you receive a high surrogate, cache it.
If you receive a low surrogate, the sequence is complete.

If you receive a high surrogate followed by a NON LOW SURROGATE, the sequence is invalid and should be discarded.
If you receive a low surrogate WITHOUT a high surrogate, the sequence is invalid and should be discarded.

Reply:
When we copy and paste (🕴) then we get the events in this order.

1) UnicodeChar = 0
     bKeyDown = 1
     wVirtualKeyCode = 18
2)  UnicodeChar = 55357
     bKeyDown = 0
     wVirtualKeyCode = 18
3)  UnicodeChar = 0
     bKeyDown = 1
     wVirtualKeyCode = 18
4)  UnicodeChar =56692
     bKeyDown = 0
     wVirtualKeyCode = 18
5) UnicodeChar = 0
     bKeyDown = 1
     wVirtualKeyCode = 18

Having said that the correct code is doing right.

Reply:
Yes, your code works fine in the _valid input_ case. If the user's clipboard contains a set of incorrect surrogate pairs (which can happen if an application is handling UTF-16 input wrong), this will incorrectly store them and parse them as though they were valid characters. UTF-16 is not a fun encoding to deal with, so it has problems like this, where applications need to understand where the encoding can go wrong.

Reply:
You can ignore this request for this pull request, but I am going to file a followup work item to explain it. Does that sound good?

Reply:
Sure, please create a work item with repro steps.